### ISSUE TITLE
explicit_bzero() causing havoc

### ISSUE THREAD

**Issue by slimhazard** [varnishcache/varnish-cache#3051]
tl;dr: I suggest that the project get rid of using the non-standard ``explicit_bzero()`` for ``ZERO_OBJ()``.

Longish rant follows.

With newer compiler versions, I'm finding that VMOD code has suddenly begun failing to compile with ``implicit declaration of function 'explicit_bzero'``. It happened on my dev laptop after updating to Debian buster (with both gcc and clang). A colleague also ran into the problem while building on a relatively recent version of Ubuntu.

It appears that newer versions of the include files have changed with respect to the feature test macros that determine what is and is not defined, depending on the C standard, POSIX version, GNU source,  BSD source, XOPEN source, etc etc. ``/usr/include/string.h`` on my machine currently has this:

```
#ifdef __USE_MISC
# include <strings.h>

/* Set N bytes of S to 0.  The compiler will not delete a call to this
   function, even if S is dead after the call.  */
extern void explicit_bzero (void *__s, size_t __n) __THROW __nonnull ((1));

/* ... other stuff ... */
#endif
```

``__USE_MISC`` in turn is defined if ``_DEFAULT_SOURCE`` is defined, or ``_GNU_SOURCE``, or a few other things. This is with:

```
$ ldd --version 
ldd (Debian GLIBC 2.28-10) 2.28
```
I've found that ``#define _DEFAULT_SOURCE 1`` makes the error go away, or even:
```
#define __USE_MISC 1
#include <string.h>
#undef __USE_MISC
```
But of course that is intolerably implementation-specific. Frankly, my current workaround is:
```
/* after including miniobj.h, or anything that includes it */
#undef ZERO_OBJ
#define ZERO_OBJ(to, sz) (void)memset((to), 0, (sz))
```
The autoconf macro ``AC_CHECK_FUNCS`` apparently runs with some sort of setting as to the source features, but it's not transparent at all what those settings might be. I suppose I could go digging through the m4 source to figure it out, but having to consider that at all is one of the bad signs.

The check for ``explicit_bzero()`` happens by way of the ``VARNISH_VMODS`` macro in ``varnish.m4``. So then anyone who uses that, and then includes ``config.h`` before ``miniobj.h`` will get the result of whatever ``AC_CHECK_FUNCS`` thought it found out about defining ``ZERO_OBJ``. ``ZERO_OBJ``, in turn, is called by ``FREE_OBJ``, and that is very commonly used for VMOD object finalization.

I suspect that there's a silent assumption at work, to the effect that everyone cargo-cultishly assumes ``_GNU_SOURCE``. (varnish-cache has ``AC_GNU_SOURCE`` in  ``configure.ac``.) I specifically *don't* want to do that in my own code if at all possible. No doubt it's an easy way to make the trouble go away, but IMO it presents too many risks about portability and compatibility. I don't want my code to become unwittingly dependent on GNU features or any other non-standard features. And IMO "just say ``GNU_SOURCE``" is not at all compatible with the cross-platform goals of the Varnish project.

If I turn on ``_DEFAULT_SOURCE`` in my code, I have no idea what other effects it might have. Unless I go studying the glibc manual, but if it's just to get ``explicit_bzero`` to stop causing trouble, then it's causing too much trouble.

Remember that just a few years ago, most of us didn't think about alpine and the musl C library. A few years before that, we didn't think about clang. New C libs and compilers are a good thing, so it's just not the case, or won't always be, that "all the world is ``_GNU_SOURCE``", and IMO we shouldn't want that.

Above all, IMO having ``explicit_bzero`` for ``ZERO_OBJ`` is not so much of an advantage that it warrants all the trouble it causes.

I see the point in the added safety of calling ``ZERO_OBJ`` in ``FREE_OBJ`` -- clear the struct before freeing it, ensuring that the value at the ``.magic`` is really 0, then an improper access will blow up twice as hard on the next assertion (understandable in view of recent events). For that, ``memset`` preferably shouldn't be optimized away. But the scenario is after all a case of pointer use after free, which ought to blow up for that reason alone.

At any rate, IMO there must be a better solution for that than forcing everyone into ``_GNU_SOURCE``, or ``_DEFAULT_SOURCE`` or anything else with so many far-reaching consequences. Someday when the bambusl C library and the kerblam compiler are better than anything we ever imagined, we'll be glad that we didn't box ourselves in.

**Comment by bsdphk** [comment]
I just _have_ to point out that some of us have insisted on supporing non-mainstream operatingsystems and compilers from day one :-)

**Comment by nigoroll** [comment]
I do not understand how the previous comment relates to the initial one

**Comment by slimhazard** [comment]
@nigoroll, at the risk of getting repetitive (which is making this feel similar to retelling a botched joke), the initial comment speculates that this sort of thing has gone unnoticed because of the silent assumption that everyone switches on ``_GNU_SOURCE``. (The initial comment references cargo cults in so doing.) And it anticipates a reaction to the effect of "quit complaining and just turn on ``_GNU_SOURCE`` like everyone else".

My comment goes on (excessively?) about what's wrong with that mindset, above all that it is hostile to the goals of platform and libc independence.

So I take @bsdphk's response as supporting at least part of the rant, and re-iterating that those are indeed project goals.

The meta-message goes beyond just the thing with ``explicit_bzero``. Varnish provides APIs for other developers (VRT, VUT, the other support for VMODs, etc), and IMO should not make assumptions about the developers' code, silent or otherwise, beyond C and POSIX standards. Certainly not non-standard feature sets like GNU, and devs certainly shouldn't be *forced* into something like that.

**Comment by bsdphk** [comment]
Solved in #3057

**Comment by slimhazard** [comment]
On 9/16/19 13:37, Poul-Henning Kamp wrote:
> Solved in #3057

Repeating my comment in github on the commit (to make sure that everyone
can see it):

The EXPLICIT_BZERO check is still AC_REQUIRE'd in varnish.m4, from
VARNISH_PREREQ and _VARNISH_CHECK_DEVEL, which are used in VMOD
development. Since it's not defined now, this leads to a cascade of
error messages when autogen.sh is called for a VMOD.

Apparently these can be ignored -- I can now compile without a
workaround for ZERO_OBJ. Thank you, I will stop screaming into the abyss
now.

It's worth pointing out, however, that Colin Percival concluded in his
blog that not even this solution guarantees that the memset call won't
be optimized out:

https://www.daemonology.net/blog/2014-09-05-erratum.html

But this gets us much closer to something that will work on most
platforms. It's OpenSSL's solution for wiping keys in memory, so one
hopes that it works most of the time.


-- 
** * * UPLEX - Nils Goroll Systemoptimierung

Scheffelstraße 32
22301 Hamburg

Tel +49 40 2880 5731
Mob +49 176 636 90917
Fax +49 40 42949753

http://uplex.de

**Comment by Dridi** [comment]
While I understand your frustration here, I am myself not too happy that your biased bug report led to a dissatisfying fix for the both of us. If I understand correctly the fix was incomplete in the 6.3 release and addressed in #3062.

The problem you had with `explicit_bzero(3)` was that detection was purely based on the shared object, regardless of its availability at the source level. We should have fixed that instead of removing support altogether. And in the absence of `explicit_bzero()` we could have a fallback to the current workaround.
